diff --git a/lv_conf.h b/lv_conf.h
index f33f052..47f8191 100644
--- a/lv_conf.h
+++ b/lv_conf.h
@@ -430,10 +430,10 @@ typedef void * lv_font_user_data_t;
 #define LV_THEME_DEFAULT_COLOR_PRIMARY      lv_color_hex(0x01a2b1)
 #define LV_THEME_DEFAULT_COLOR_SECONDARY    lv_color_hex(0x44d1b6)
 #define LV_THEME_DEFAULT_FLAG               LV_THEME_MATERIAL_FLAG_LIGHT
-#define LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_12
-#define LV_THEME_DEFAULT_FONT_NORMAL        &lv_font_montserrat_16
-#define LV_THEME_DEFAULT_FONT_SUBTITLE      &lv_font_montserrat_20
-#define LV_THEME_DEFAULT_FONT_TITLE         &lv_font_montserrat_24
+#define LV_THEME_DEFAULT_FONT_SMALL         ( (LV_HOR_RES_MAX>300) ? (&lv_font_montserrat_36) : (&lv_font_montserrat_12))
+#define LV_THEME_DEFAULT_FONT_NORMAL        ( (LV_HOR_RES_MAX>300) ? (&lv_font_montserrat_40) : (&lv_font_montserrat_16))
+#define LV_THEME_DEFAULT_FONT_SUBTITLE      ( (LV_HOR_RES_MAX>300) ? (&lv_font_montserrat_44) : (&lv_font_montserrat_20))
+#define LV_THEME_DEFAULT_FONT_TITLE         ( (LV_HOR_RES_MAX>300) ? (&lv_font_montserrat_48) : (&lv_font_montserrat_24))
 
 /*=================
  *  Text settings
