diff --git a/lvgl.mk b/lvgl.mk
index bbea98d3..11b6e5cd 100755
--- a/lvgl.mk
+++ b/lvgl.mk
@@ -6,5 +6,6 @@ include $(LVGL_DIR)/$(LVGL_DIR_NAME)/src/lv_misc/lv_misc.mk
 include $(LVGL_DIR)/$(LVGL_DIR_NAME)/src/lv_themes/lv_themes.mk
 include $(LVGL_DIR)/$(LVGL_DIR_NAME)/src/lv_draw/lv_draw.mk
 include $(LVGL_DIR)/$(LVGL_DIR_NAME)/src/lv_gpu/lv_gpu.mk
+include $(LVGL_DIR)/$(LVGL_DIR_NAME)/src/extra/extra.mk
 
 
