diff --git a/src/lv_conf_internal.h b/src/lv_conf_internal.h
index 90e33f69..29b62d27 100755
--- a/src/lv_conf_internal.h
+++ b/src/lv_conf_internal.h
@@ -1117,6 +1117,34 @@ e.g. "stm32f769xx.h" or "stm32f429xx.h" */
 #define LV_USE_WIN      1
 #endif
 
+/*=====================
+ * 3rd party libraries
+ *=====================*/
+
+/*FreeType library*/
+#ifndef LV_USE_FREETYPE
+#define LV_USE_FREETYPE      1
+#endif
+#if LV_USE_FREETYPE
+    #ifndef LV_USE_FT_CACHE_MANAGER
+    #  define LV_USE_FT_CACHE_MANAGER 1
+    #endif
+    #if LV_USE_FT_CACHE_MANAGER >= 0
+        /*Memory used by FreeType to cache characters [bytes] (-1: no caching)*/
+        #ifndef LV_FREETYPE_CACHE_SIZE
+        #  define LV_FREETYPE_CACHE_SIZE (16 * 1024)
+        #endif
+        /* Maximum number of opened FT_Face/FT_Size objects managed by this cache instance. */
+        /* (0:use system defaults) */
+        #ifndef LV_FREETYPE_CACHE_FT_FACES
+        #  define LV_FREETYPE_CACHE_FT_FACES 0
+        #endif
+        #ifndef LV_FREETYPE_CACHE_FT_SIZES
+        #  define LV_FREETYPE_CACHE_FT_SIZES 0
+        #endif
+    #endif
+#endif
+
 /*==================
  * Non-user section
  *==================*/
