diff --git a/src/extra/lv_extra.h b/src/extra/lv_extra.h
new file mode 100755
index 00000000..ea9ce381
--- /dev/null
+++ b/src/extra/lv_extra.h
@@ -0,0 +1,42 @@
+/**
+ * @file lv_extra.h
+ *
+ */
+
+#ifndef LV_EXTRA_H
+#define LV_EXTRA_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+/*********************
+ *      DEFINES
+ *********************/
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+/**********************
+ * GLOBAL PROTOTYPES
+ **********************/
+
+/**
+ * Initialize the extra components
+ */
+void lv_extra_init(void);
+
+/**********************
+ *      MACROS
+ **********************/
+
+#ifdef __cplusplus
+} /*extern "C"*/
+#endif
+
+#endif /*LV_EXTRA_H*/
