From c990228295c5864db8fa20820c100cf42904f846 Mon Sep 17 00:00:00 2001
From: guobin <guobin.xue@unisoc.com>
Date: Sat, 19 Jan 2019 15:37:34 +0800
Subject: [PATCH 1/5] Sp9832a fix build error for yocto

---
 include/configs/sp9832a_2h11_volte.h | 70 +++++++++++++++++++-----------------
 1 file changed, 37 insertions(+), 33 deletions(-)

diff --git a/include/configs/sp9832a_2h11_volte.h b/include/configs/sp9832a_2h11_volte.h
index b1cd8c6..c3aa721 100644
--- a/include/configs/sp9832a_2h11_volte.h
+++ b/include/configs/sp9832a_2h11_volte.h
@@ -24,8 +24,8 @@

 #include "sprd_sharkl_modem_volte.h"

-// #define CONFIG_SECURE_BOOT
-// #define CONFIG_ROM_VERIFY_SPL
+/* #define CONFIG_SECURE_BOOT */
+/* #define CONFIG_ROM_VERIFY_SPL */

 #define PRIMPUKPATH "/dev/block/mmcblk0boot0"
 #define PRIMPUKSTART 512
@@ -38,7 +38,7 @@
 #define DT_HARDWARE_ID 1
 #define DT_SOC_VER     0x20000

-//only used in fdl2 .in uart download, the debug infors  from  serial will break the download process.
+/* only used in fdl2 .in uart download, the debug infors  from  serial will break the download process. */

 #define BOOT_NATIVE_LINUX_MODEM  1
 #define CONFIG_SILENT_CONSOLE
@@ -66,7 +66,7 @@
 #define CONFIG_SP9832A_2H11_VOLTE
 #define CONFIG_SC9630
 #define CONFIG_ADIE_SC2723
-//#define CONFIG_ARCH_SCX35L
+/* #define CONFIG_ARCH_SCX35L */
 #define CONFIG_SPL_32K


@@ -143,7 +143,7 @@
 #endif


-//The macro deffined in config.mk of uboot 
+/* The macro deffined in config.mk of uboot */
 #ifdef SP9832A_2H11_32V4_VOLTE_LCD
 #define  CONFIG_FB_LOW_RES_SIMU
 #endif
@@ -203,15 +203,18 @@
 */

 /* DDR */
+/*
 //#define DDR_CLK 464
 //---these three macro below,only one can be open
 //#define DDR_LPDDR1
 //#define DDR_LPDDR2
 //#define DDR_DDR3
+*/

 #define CONFIG_DDR_AUTO_DETECT
 #define CONFIG_NR_DRAM_BANKS_ADDR_IN_IRAM    0x1C00   /* IRAM store ddr info */

+/*
 //#define DDR_TYPE DRAM_LPDDR2_2CS_8G_X32
 //#define DDR_TYPE DRAM_LPDDR2_1CS_4G_X32
 //#define DDR_TYPE DRAM_LPDDR2_1CS_8G_X32
@@ -231,10 +234,11 @@
 //#define PUBL_LPDDR1_DS PUBL_LPDDR1_DS_48OHM
 //#define PUBL_LPDDR2_DS PUBL_LPDDR2_DS_40OHM
 //#define PUBL_DDR3_DS   PUBL_DDR3_DS_34OHM
+*/

 /* NAND */
 #define CONFIG_NAND_SC9630
-//#define CONFIG_SPRD_NAND_REGS_BASE	(0x21100000)
+/* #define CONFIG_SPRD_NAND_REGS_BASE	(0x21100000) */
 #define CONFIG_SYS_MAX_NAND_DEVICE	1
 #define CONFIG_SYS_NAND_BASE		(0x21100000)

@@ -288,11 +292,11 @@
 #define CONFIG_DSIH_VERSION_1P21A
 #define CONFIG_SPLASH_SCREEN
 #define LCD_BPP LCD_COLOR16
-//#define CONFIG_LCD_FWVGA
+/* #define CONFIG_LCD_FWVGA */
 #define CONFIG_LCD_720P
 #define CONFIG_CMD_BMP

-//#define CONFIG_FB_LCD_OTM8019A_MIPI
+/* #define CONFIG_FB_LCD_OTM8019A_MIPI */
 #define CONFIG_FB_LCD_HX8394A_MIPI
 #define CONFIG_FB_LCD_ILI9881C_MIPI
 #ifdef  CONFIG_FB_LOW_RES_SIMU
@@ -308,32 +312,32 @@
 #define CONSOLE_COLOR_MAGENTA 0x001f
 #define CONSOLE_COLOR_CYAN 0x001f
 #endif
-#endif // CONFIG_LCD
+#endif /* CONFIG_LCD */

-/*for sysdump*/
+/* for sysdump */
 #define CONFIG_FS_FAT
 #define CONFIG_SPRD_SYSDUMP
 #define CONFIG_RAMDUMP_NO_SPLIT 1	/* Don't split sysdump file */
-#define REAL_SDRAM_SIZE 0x40000000	/*dump 1G */
+#define REAL_SDRAM_SIZE 0x40000000	/* dump 1G */


 #define CALIBRATE_ENUM_MS 3000
 #define CALIBRATE_IO_MS 2000

-//#define LOW_BAT_ADC_LEVEL 782 /*phone battery adc value low than this value will not boot up*/
+/* #define LOW_BAT_ADC_LEVEL 782 */ /*phone battery adc value low than this value will not boot up*/
 #define LOW_BAT_VOL            3400 /*phone battery voltage low than this value will not boot up*/
-#define LOW_BAT_VOL_CHG        3300    //3.3V charger connect
+#define LOW_BAT_VOL_CHG        3300    /* 3.3V charger connect */

-#define PWR_KEY_DETECT_CNT 2 /*this should match the count of boot_pwr_check() function */
+#define PWR_KEY_DETECT_CNT 2 /* this should match the count of boot_pwr_check() function */
 #define ALARM_LEAD_SET_MS 0 /* time set for alarm boot in advancd */


-/*rf board id */
+/* rf board id */
 #define RF_BAND_INFO
 #ifdef RF_BAND_INFO
 #define LB_GPIO_NUM   134
 #define MB_GPIO_NUM   135
-//#define HB_GPIO_NUM  0
+/* #define HB_GPIO_NUM  0 */
 #define ADC_CHANNEL_FOR_NV  1
 #endif

@@ -348,13 +352,13 @@


 #define CONFIG_PBINT_7S_RESET_V1
-/*7S reset config*/
-#define CONFIG_7S_RST_MODULE_EN		1	//0:disable module; 1:enable module
+/* 7S reset config */
+#define CONFIG_7S_RST_MODULE_EN		1	/* 0:disable module; 1:enable module */

-#define CONFIG_7S_RST_SW_MODE		1	//0:hw reset,1:arm reset,power keep on
-#define CONFIG_7S_RST_SHORT_MODE	1	//0:long press then release key to trigger;1:press key some time to trigger
-#define CONFIG_7S_RST_2KEY_MODE		0	//0:1Key--Normal mode; 1:2KEY
-#define CONFIG_7S_RST_THRESHOLD		7	//7S, hold key down for this time to trigger
+#define CONFIG_7S_RST_SW_MODE		1	/* 0:hw reset,1:arm reset,power keep on */
+#define CONFIG_7S_RST_SHORT_MODE	1	/* 0:long press then release key to trigger;1:press key some time to trigger */
+#define CONFIG_7S_RST_2KEY_MODE		0	/* 0:1Key--Normal mode; 1:2KEY */
+#define CONFIG_7S_RST_THRESHOLD		7	/* 7S, hold key down for this time to trigger */

 #define DT_ADR          0x85400000
 #define KERNEL_ADR      0x80008000
@@ -370,36 +374,36 @@



-/*rf board id */
+/* rf board id */

 #define RF_BAND_INFO

 #ifdef RF_BAND_INFO
 #define LB_GPIO_NUM 			134
 #define MB_GPIO_NUM 			135
-//#define HB_GPIO_NUM 			0
-//#define ADC_CHANNEL_FOR_NV		0
+/* #define HB_GPIO_NUM 			0 */
+/* #define ADC_CHANNEL_FOR_NV		0 */
 #endif


 /* Chip Driver Macro Definitions Start*/
-/*sprd adc*/
+/* sprd adc */
 #define CONFIG_SPRD_ADC

-/*sprd adi*/
+/* sprd adi */
 #define CONFIG_SPRD_ADI

-/*sprd gpio*/
+/* sprd gpio */
 #define CONFIG_SPRD_GPIO

-/*sprd rtc*/
+/* sprd rtc */
 #define CONFIG_RTC_SPRD

-/*sprd watchdog*/
+/* sprd watchdog */
 #define CONFIG_HW_WATCHDOG
 #define CONFIG_SPRD_WATCHDOG

- /*Serial Info*/
+/* Serial Info */
 #define CONFIG_SPRD_UART		1
 #define CONFIG_SYS_SC8800X_UART1	1
 #define CONFIG_CONS_INDEX	1	/* use UART0 for console */
@@ -410,8 +414,8 @@
 #define CONFIG_SYS_SERIAL0		0x70000000
 #define CONFIG_SYS_SERIAL1		0x70100000

-/* Chip Driver Macro Definitions End*/
-/*active arm7 ram before access to it*/
+/* Chip Driver Macro Definitions End */
+/* active arm7 ram before access to it */
 #define CONFIG_ARM7_RAM_ACTIVE

 #define CONFIG_SPRD_EXT_IC_POWER
--
1.9.1
