From f4e4814164dc72688682249a7e35a732ff020e86 Mon Sep 17 00:00:00 2001
From: guobin <guobin.xue@unisoc.com>
Date: Mon, 21 Jan 2019 13:36:57 +0800
Subject: [PATCH 4/5] Sp9832a yocto: fixed sp9832a_2h11_volte build warning

---
 include/configs/sp9832a_2h11_volte.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/sp9832a_2h11_volte.h b/include/configs/sp9832a_2h11_volte.h
index c3aa721..89ac30c 100644
--- a/include/configs/sp9832a_2h11_volte.h
+++ b/include/configs/sp9832a_2h11_volte.h
@@ -92,7 +92,7 @@

 #define CONFIG_FS_FAT
 #define CONFIG_UEFI_PARTITION
-#define CONFIG_EFI_PARTITION
+/* #define CONFIG_EFI_PARTITION */
 #define CONFIG_EXT4_SPARSE_DOWNLOAD
 #endif

--
1.9.1
