From ab2be290850113814c931cf9f9982073b17fd46d Mon Sep 17 00:00:00 2001
From: fujia <fujia@bingtangtech.com>
Date: Thu, 17 Aug 2023 15:30:05 +0800
Subject: [PATCH] add key okey value

---
 include/uapi/linux/input-event-codes.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/uapi/linux/input-event-codes.h b/include/uapi/linux/input-event-codes.h
index 311a57f3e01a..60f241532579 100644
--- a/include/uapi/linux/input-event-codes.h
+++ b/include/uapi/linux/input-event-codes.h
@@ -336,6 +336,7 @@
 #define KEY_RFKILL		247	/* Key that controls all radios */
 
 #define KEY_MICMUTE		248	/* Mute / unmute the microphone */
+#define KEY_OKAY        249
 
 /* Code 255 is reserved for special needs of AT keyboard driver */
 
-- 
2.25.1

