diff --git a/configure.ac b/configure.ac
index 156e5c3..f7639c8 100755
--- a/configure.ac
+++ b/configure.ac
@@ -1314,7 +1314,9 @@ AS_IF([test "x$HAVE_SYSTEMD_JOURNAL" = "x1"], AC_DEFINE([HAVE_SYSTEMD_JOURNAL],
 AC_ARG_ENABLE([whalehal],
     AS_HELP_STRING([--enable-whalehal],[Use whale hal for audio]))
 
+AC_SUBST(USE_WHALE_HAL)
 AM_CONDITIONAL([USE_WHALE_HAL], [test "x$enable_whalehal" = "xyes"])
+AS_IF([test "x$enable_whalehal" = "xyes"], USE_WHALE_HAL=1, USE_WHALE_HAL=0)
 
 #### Build and Install man pages ####
 
diff --git a/src/daemon/daemon.conf.in b/src/daemon/daemon.conf.in
index da437b7..b7a9357 100755
--- a/src/daemon/daemon.conf.in
+++ b/src/daemon/daemon.conf.in
@@ -81,8 +81,14 @@ ifelse(@HAVE_SYS_RESOURCE_H@, 1, [dnl
 ])dnl
 
 ; default-sample-format = s16le
-; default-sample-rate = 44100
+ifelse(@USE_WHALE_HAL@, 1, [dnl
+default-sample-rate = 48000
+alternate-sample-rate = 48000
+])dnl
+ifelse(@USE_WHALE_HAL@, 0, [dnl
+default-sample-rate = 44100
 alternate-sample-rate = 44100
+])dnl
 ; default-sample-channels = 2
 ; default-channel-map = front-left,front-right
 
